src/analyses/goto_check.cpp

```diff
@@ -351,6 +365,53 @@ void goto_checkt::mod_by_zero_check(
     guard);
 }
 
+/// Check if the frontend has flagged a given expression for range checking.
+/// \param e: The expression that is to be checked.
+/// \return: True or false based on whether the flag was set or not.
+static bool range_check_flag_set(const exprt &e)
+{
+  return e.get_bool("range_check");
+}
+
+/// Add a new range check on an arithmetic expression in the list
+/// of assertions.
+/// \param expr: The arithmetic expression
+/// \param guard: A guard expression.
+void goto_checkt::ranged_type_check(const exprt &expr, const guardt &guard)
+{
+  if(!range_check_flag_set(expr))
+    return;
+
+  const typet &real_type = ns.follow(expr.type());
+  if(real_type.id() != ID_signedbv)
+    return;
+  const exprt &lower_bound =
```

Comment:
We actually have a ranged type! Look up range_typet.

Reply:
The `range_typet` is not appropriate to use at this point in time, because the frontend `gnat2goto` is producing `signed_bv` values that also have a bound value bolted on (see https://github.com/diffblue/gnat2goto/blob/master/gnat2goto/driver/tree_walk.adb#L2804). This will mean that for this to change, we have to also make a significant change to `gnat2goto`. As this work is for getting preliminary support for `gnat2goto` in `cbmc` I suggest this stays as it is and gets updated later. What do you think about this @kroening ?

Reply:
I think it makes more sense to keep these as `signedbv`, as in Ada intermediary expressions are allowed to escape range boundaries as long as they stay within the boundaries of the underlying implementation (i.e. no under/overflow) and end up within the right range during the final assignment.